core/dbt/parser/schemas.py

```diff
@@ -67,109 +67,160 @@ def as_kwarg(key, value):
-        **{
-            'model': model['name'],
-            'macro': macro_name,
-            'kwargs': ", ".join(kwargs)
-        }
+    """
+    TEST_NAME_PATTERN = re.compile(
+        r'((?P<test_namespace>([a-zA-Z_][0-9a-zA-Z_]*))\.)?'
+        r'(?P<test_name>([a-zA-Z_][0-9a-zA-Z_]*))'
+        r'(?P<test_modifiers>(, [a-zA-Z_][0-9a-zA-Z_]*=[^,=\s]*)*)'
     )
-    return raw_sql
-
 
-def build_source_test_raw_sql(test_namespace, source, table, test_type,
-                              test_args):
-    """Build the raw SQL from a source test definition.
+    def __init__(self, test, target, column_name, package_name):
+        test_name, test_args = self.build_test_args(test, column_name)
+        self.args = test_args
+        self.package_name = package_name
+        self.target = target
 
-    :param test_namespace: The test's namespace, if one exists
-    :param source: The source under test.
-    :param table: The table under test
-    :param test_type: The type of the test (unique_id, etc)
-    :param test_args: The arguments passed to the test as a list of `key=value`
-        strings
-    :return: A string of raw sql for the test node.
-    """
-    # sort the dict so the keys are rendered deterministically (for tests)
-    kwargs = [as_kwarg(key, test_args[key]) for key in sorted(test_args)]
-
-    if test_namespace is None:
-        macro_name = "test_{}".format(test_type)
-    else:
-        macro_name = "{}.test_{}".format(test_namespace, test_type)
-
-    raw_sql = (
-        "{{{{ {macro}(model=source('{source}', '{table}'), {kwargs}) }}}}"
-        .format(
-            source=source['name'],
-            table=table['name'],
-            macro=macro_name,
-            kwargs=", ".join(kwargs))
-    )
-    return raw_sql
+        match = self.TEST_NAME_PATTERN.match(test_name)
+        if match is None:
+            dbt.exceptions.raise_compiler_error(
+                'Test name string did not match expected pattern: {}'
+                .format(test_name)
+            )
 
+        groups = match.groupdict()
+        self.name = groups['test_name']
+        self.namespace = groups['test_namespace']
+        self.modifiers = self.build_test_modifiers(groups['test_modifiers'])
 
-def calculate_test_namespace(test_type, package_name):
-    test_namespace = None
-    split = test_type.split('.')
-    if len(split) > 1:
-        test_type = split[1]
-        package_name = split[0]
-        test_namespace = package_name
+        if self.namespace is not None:
+            self.package_name = self.namespace
 
-    return test_namespace, test_type, package_name
+    @staticmethod
+    def build_test_modifiers(modifiers_value):
+        modifiers = {}
+        if modifiers_value is None:
+            return modifiers
+
+        modifiers_value = modifiers_value.strip(', \t')
+        if not modifiers_value:
+            return modifiers
+
+        for entry in modifiers_value.split(','):
+            if '=' not in entry:
```

Comment:
I don't think this works as expected, the following spec compiles and runs whereas I would expect to see a compiler error:
```
models:
    - name: my_model
      columns:
          - name: id
            tests:
                - unique, invalid
```